src/Microsoft.Extensions.HealthChecks/HealthCheckService.cs

```diff
@@ -0,0 +1,98 @@
+
+        /// <summary>
+        /// A <see cref="IReadOnlyDictionary{TKey, T}"/> containing all the health checks registered in the application.
+        /// </summary>
+        /// <remarks>
+        /// The key maps to the <see cref="IHealthCheck.Name"/> property of the health check, and the value is the <see cref="IHealthCheck"/>
+        /// instance itself.
+        /// </remarks>
+        public IReadOnlyDictionary<string, IHealthCheck> Checks { get; }
+
+        /// <summary>
+        /// Constructs a <see cref="HealthCheckService"/> from the provided collection of <see cref="IHealthCheck"/> instances.
+        /// </summary>
+        /// <param name="healthChecks">The <see cref="IHealthCheck"/> instances that have been registered in the application.</param>
+        public HealthCheckService(IEnumerable<IHealthCheck> healthChecks) : this(healthChecks, NullLogger<HealthCheckService>.Instance) { }
+
+        /// <summary>
+        /// Constructs a <see cref="HealthCheckService"/> from the provided collection of <see cref="IHealthCheck"/> instances, and the provided logger.
+        /// </summary>
+        /// <param name="healthChecks">The <see cref="IHealthCheck"/> instances that have been registered in the application.</param>
+        /// <param name="logger">A <see cref="ILogger{T}"/> that can be used to log events that occur during health check operations.</param>
+        public HealthCheckService(IEnumerable<IHealthCheck> healthChecks, ILogger<HealthCheckService> logger)
+        {
+            Checks = healthChecks.ToDictionary(c => c.Name);
+            _logger = logger;
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                foreach (var check in Checks)
+                {
+                    _logger.LogDebug("Health check '{healthCheckName}' has been registered", check.Key);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Runs all the health checks in the application and returns the aggregated status.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> which can be used to cancel the health checks</param>
+        /// <returns>
+        /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
+        /// yielding a <see cref="CompositeHealthCheckResult"/> containing the results.
+        /// </returns>
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default)
+        {
+            var results = new Dictionary<string, HealthCheckResult>(Checks.Count);
+            foreach (var pair in Checks)
+            {
+                // If the health check does things like make Database queries using EF or backend HTTP calls,
+                // it may be valuable to know that logs it generates are part of a health check. So we start a scope.
+                using (_logger.BeginScope(new HealthCheckLogScope(pair.Key)))
```

Comment:
I don't think a scope outside makes much sense. There's no data to apply to the scope outside the foreach. The scope has to live inside the loop because it's tagging each scope with the health check name.